### ISSUE TITLE
workers crashing

### ISSUE THREAD

**Issue by tetlika** [slact/nchan#649]
we use nginx 1.22.0 with nchan 1.3.4

from time to time - e.g. once per several months we have following issue -  nginx workers are crashing heavily and master process is restarting them and they crash again and again; nginx restart helps for some time

in logs we have smth like this:
`2022/09/19 23:33:56 [notice] 16030#0: start worker process 3679

2022/09/19 23:33:56 [notice] 16030#0: signal 29 (SIGIO) received

2022/09/19 23:33:56 [notice] 16030#0: signal 17 (SIGCHLD) received from 3677

2022/09/19 23:33:56 [alert] 16030#0: worker process 3677 exited on signal 11

2022/09/19 23:33:56 [notice] 16030#0: start worker process 3680

2022/09/19 23:33:56 [notice] 16030#0: signal 29 (SIGIO) received

2022/09/19 23:33:56 [notice] 16030#0: signal 17 (SIGCHLD) received from 3678

2022/09/19 23:33:56 [alert] 16030#0: worker process 3678 exited on signal 11

2022/09/19 23:33:56 [notice] 16030#0: start worker process 3681

2022/09/19 23:33:56 [notice] 16030#0: signal 29 (SIGIO) received

2022/09/19 23:33:56 [notice] 16030#0: signal 17 (SIGCHLD) received from 3679

2022/09/19 23:33:56 [alert] 16030#0: worker process 3679 exited on signal 11

2022/09/19 23:33:56 [notice] 16030#0: start worker process 3682

2022/09/19 23:33:56 [notice] 16030#0: signal 29 (SIGIO) received

2022/09/19 23:33:56 [notice] 16030#0: signal 17 (SIGCHLD) received from 3680

2022/09/19 23:33:56 [alert] 16030#0: worker process 3680 exited on signal 11`

our nginx config is:
 server {
        listen 80;
        listen 443 ssl;
        ssl_session_timeout 5m;
        ssl_protocols TLSv1.2;
        ssl_certificate     /var/certs/file.crt;
        ssl_certificate_key /var/certs/file.key;
        ssl_session_cache shared:SSL:10m;
      
        location /push/pub {
          nchan_publisher http;
          nchan_channel_id $arg_id;
          allow 172.16.0.0/12;
          allow 127.0.0.1;
         deny all;
       }

       location ~ /push/ws/(.*) {
         nchan_subscriber                            websocket;
         nchan_channel_id                            $1;
         nchan_websocket_ping_interval               10s;
       }

      }

      nchan_max_reserved_memory    64M;
      nchan_message_timeout        600;  # 10 minutes
}

**Comment by tetlika** [comment]
anyone anything? :)

**Comment by dtcooper** [comment]
Having the same problem on nginx 1.23.1 - reverting to nchan versions 1.3.0 helps a bit. The workers still crash but not repeatedly.

**Comment by slact** [comment]
can you provide me with a coredump or a stack trace from the cores?

**Comment by dtcooper** [comment]
@slact Unsure exactly how to get you that. I can provide a Dockerfile I used to nchan as an nginx module?

**Comment by tetlika** [comment]
@slact how we can get coredump can you advise please? thank you

**Comment by kapouer** [comment]
@slact here's a gziped coredump made on debian/sid with nginx 1.22.1 and nchan 1.3.5.

[nginx-nchan.dump.gz](https://github.com/slact/nchan/files/10238047/nginx-nchan.dump.gz)

The text output was

```
           PID: 1151735 (nginx)
           UID: 33 (www-data)
           GID: 33 (www-data)
        Signal: 11 (SEGV)
     Timestamp: Thu 2022-12-15 15:47:42 CET (1min 21s ago)
  Command Line: $'nginx: master process nginx -g daemon off; master_process on;'
    Executable: /usr/sbin/nginx
 Control Group: /user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-8124bcb0-5341-44c5-a908-9e7e7c8d7f7c.scope
          Unit: user@1000.service
     User Unit: vte-spawn-8124bcb0-5341-44c5-a908-9e7e7c8d7f7c.scope
         Slice: user-1000.slice
     Owner UID: 1000 (dev)
       Boot ID: 496e1a92ab0c4767b5a4ea675b00bae6
    Machine ID: 12e11ceb84fefe701a02ef52000007db
      Hostname: lal.localdomain
       Storage: /var/lib/systemd/coredump/core.nginx.33.496e1a92ab0c4767b5a4ea675b00bae6.1151735.1671115662000000.zst (present)
  Size on Disk: 253.3K
       Message: Process 1151735 (nginx) of user 33 dumped core.
                
                Module libnss_systemd.so.2 from deb systemd-252.3-2.amd64
                Stack trace of thread 1151735:
                #0  0x00007f0160433024 nchan_stats_init_worker (ngx_nchan_module.so + 0x33024)
                #1  0x00007f016041cb28 n/a (ngx_nchan_module.so + 0x1cb28)
                #2  0x000055cc5c38e3fe n/a (nginx + 0x503fe)
                #3  0x000055cc5c38e90d n/a (nginx + 0x5090d)
                #4  0x000055cc5c38cff3 ngx_spawn_process (nginx + 0x4eff3)
                #5  0x000055cc5c38e178 n/a (nginx + 0x50178)
                #6  0x000055cc5c38faa2 ngx_master_process_cycle (nginx + 0x51aa2)
                #7  0x000055cc5c36230e main (nginx + 0x2430e)
                #8  0x00007f0160a4618a n/a (libc.so.6 + 0x2718a)
                #9  0x00007f0160a46245 __libc_start_main (libc.so.6 + 0x27245)
                #10 0x000055cc5c362391 _start (nginx + 0x24391)
                ELF object binary architecture: AMD x86-64
```

**Comment by kapouer** [comment]
@dtcooper @tetlika it's really easy to get a core dump these days:
- install systemd-coredump
- crash your program
- coredumpctl dump --output crash.dump
- gzip crash.dump
- uninstall systemd-coredump (to avoid core dumps filling your storage)

**Comment by slact** [comment]
thanks @kapouer , the issue is now fixed by pr #659

**Comment by tetlika** [comment]
looks fix didnt work, I still have nginx crashing when auth is enabled, in dmesg I have:
`Wed Dec 28 09:31:54 2022] nginx[1793]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4 in nginx[400000+1be000]
[Wed Dec 28 09:31:54 2022] nginx[1796]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4 in nginx[400000+1be000]
[Wed Dec 28 09:31:56 2022] nginx[1792]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4 in nginx[400000+1be000]
[Wed Dec 28 09:31:56 2022] nginx[1798]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4 in nginx[400000+1be000]
[Wed Dec 28 09:32:00 2022] nginx[1774]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4 in nginx[400000+1be000]
[Wed Dec 28 09:32:00 2022] nginx[1797]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4 in nginx[400000+1be000]
[Wed Dec 28 09:32:07 2022] nginx[1807]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4 in nginx[400000+1be000]
[Wed Dec 28 09:32:07 2022] nginx[1802]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4 in nginx[400000+1be000]
[Wed Dec 28 09:32:10 2022] nginx[1803]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4
[Wed Dec 28 09:32:10 2022] nginx[1808]: segfault at 0 ip 000000000042d2ff sp 00007ffcff2cfc80 error 4`

in nginx log:
`2022/12/28 09:31:39 [notice] 1486#0: signal 17 (SIGCHLD) received from 1676
2022/12/28 09:31:39 [alert] 1486#0: worker process 1676 exited on signal 11
2022/12/28 09:31:39 [notice] 1486#0: start worker process 1683
2022/12/28 09:31:39 [notice] 1486#0: signal 29 (SIGIO) received
2022/12/28 09:31:39 [notice] 1486#0: signal 17 (SIGCHLD) received from 1682
2022/12/28 09:31:39 [alert] 1486#0: worker process 1682 exited on signal 11
2022/12/28 09:31:39 [notice] 1486#0: start worker process 1684
2022/12/28 09:31:39 [notice] 1486#0: signal 29 (SIGIO) received
2022/12/28 09:31:39 [notice] 1486#0: signal 17 (SIGCHLD) received from 1678
2022/12/28 09:31:39 [alert] 1486#0: worker process 1678 exited on signal 11
2022/12/28 09:31:39 [notice] 1486#0: start worker process 1685
2022/12/28 09:31:39 [notice] 1486#0: signal 29 (SIGIO) received
2022/12/28 09:31:39 [notice] 1486#0: signal 17 (SIGCHLD) received from 1681
2022/12/28 09:31:39 [alert] 1486#0: worker process 1681 exited on signal 11
2022/12/28 09:31:39 [notice] 1486#0: start worker process 1686
2022/12/28 09:31:39 [notice] 1486#0: signal 29 (SIGIO) received
2022/12/28 09:31:39 [notice] 1486#0: signal 17 (SIGCHLD) received from 1685
2022/12/28 09:31:39 [alert] 1486#0: worker process 1685 exited on signal 11
2022/12/28 09:31:39 [notice] 1486#0: start worker process 1687
2022/12/28 09:31:39 [notice] 1486#0: signal 29 (SIGIO) received
2022/12/28 09:31:39 [notice] 1486#0: signal 17 (SIGCHLD) received from 1683
2022/12/28 09:31:39 [alert] 1486#0: worker process 1683 exited on signal 11
2022/12/28 09:31:39 [notice] 1486#0: start worker process 1688
2022/12/28 09:31:39 [notice] 1486#0: signal 29 (SIGIO) received
2022/12/28 09:31:39 [notice] 1486#0: signal 17 (SIGCHLD) received from 1684
2022/12/28 09:31:39 [alert] 1486#0: worker process 1684 exited on signal 11
2022/12/28 09:31:39 [notice] 1486#0: start worker process 1689
2022/12/28 09:31:39 [notice] 1486#0: signal 29 (SIGIO) received
2022/12/28 09:31:39 [notice] 1486#0: signal 17 (SIGCHLD) received from 1687
2022/12/28 09:31:39 [alert] 1486#0: worker process 1687 exited on signal 11
2022/12/28 09:31:39 [notice] 1486#0: start worker process 1690
2022/12/28 09:31:39 [notice] 1486#0: signal 29 (SIGIO) received
2022/12/28 09:31:39 [notice] 1486#0: signal 17 (SIGCHLD) received from 1686
2022/12/28 09:31:39 [alert] 1486#0: worker process 1686 exited on signal 11
2022/12/28 09:31:39 [notice] 1486#0: start worker process 1691
2022/12/28 09:31:39 [notice] 1486#0: signal 29 (SIGIO) received`
